src/gg.regression.unity.bots/Editor/Scripts/RGLegacyInputUtility/RGLegacyInputInstrumentation.cs

```diff
@@ -260,17 +259,77 @@ public static Assembly FindRGAssembly()
 
         private static void OnAssemblyCompiled(string assemblyAssetPath, CompilerMessage[] messages)
         {
-            Assembly rgAssembly = FindRGAssembly();
-            InstrumentAssemblyIfNeeded(assemblyAssetPath, rgAssembly);
+            try
+            {
+                Assembly rgAssembly = FindRGAssembly();
+                InstrumentAssemblyIfNeeded(assemblyAssetPath, rgAssembly);
+            }
+            catch (IOException e)
+            {
+                // If the instrumentation failed here, then there is nothing we can do since this could be part of a player build process
+                RGDebug.Log($"Instrumentation of legacy input APIs failed for {assemblyAssetPath}. Simulating legacy inputs may not work, try re-building the game assemblies.\n{e.Message}\n{e.StackTrace}");
+            }
         }
 
-        private static void InstrumentExistingAssemblies()
+        static void ScheduledInstrumentationLoop()
         {
-            Assembly[] assemblies = CompilationPipeline.GetAssemblies(AssembliesType.Player);
-            Assembly rgAssembly = FindRGAssembly();
-            foreach (Assembly assembly in assemblies)
+            bool done;
+            if (EditorApplication.timeSinceStartup >= _scheduledInstrumentationTime)
             {
-                InstrumentAssemblyIfNeeded(assembly.outputPath, rgAssembly);
+                if (InstrumentExistingAssemblies())
+                {
+                    done = true;
+                }
+                else
+                {
+                    // if the instrumentation failed, attempt a couple more times with exponential backoff timeout
+                    if (_numInstrumentationAttempts < MaxAttempts-1)
+                    {
+                        double timeout = 3.0 * Math.Pow(2.0, _numInstrumentationAttempts); // 3 sec, 6 sec, 12 sec
+                        _scheduledInstrumentationTime = EditorApplication.timeSinceStartup + timeout;
+                        ++_numInstrumentationAttempts;
+                        done = false;
+                    }
+                    else
+                    {
+                        done = true;
+                    }
+                }
+            }
+            else
+            {
+                done = false;
+            }
+            if (done)
+            {
+                EditorApplication.update -= ScheduledInstrumentationLoop;
+            }
+        }
+        
+        private static bool InstrumentExistingAssemblies()
+        {
+            try
+            {
+                Assembly[] assemblies = CompilationPipeline.GetAssemblies(AssembliesType.Player);
+                Assembly rgAssembly = FindRGAssembly();
+                foreach (Assembly assembly in assemblies)
+                {
+                    InstrumentAssemblyIfNeeded(assembly.outputPath, rgAssembly);
+                }
+                return true;
```

Comment:
do we need a `_numInstrumentationAttempts = 0;` code line here before the return to reset this ?

Reply:
Technically it isn't needed since the loop can only ever happen once per static initialization, but for good measure I've added in a reset after the loop terminates:

https://github.com/Regression-Games/RGUnityBots/blob/3ab36906770a6cfa866c7bc823d03dee7fba3f81/src/gg.regression.unity.bots/Editor/Scripts/RGLegacyInputUtility/RGLegacyInputInstrumentation.cs#L303-L307